receive_buffer.go

```diff
@@ -138,6 +138,13 @@ func (buf *receiveBuffer) ackIfNecessary() {
 }
 
 func (buf *receiveBuffer) sendACK() {
+	buf.mx.RLock()
+	closed := buf.closed
+	buf.mx.RUnlock()
+	if closed {
+		// Don't bother acking
+		return
+	}
```

Comment:
Why not just:

```
	buf.mx.RLock()
 	if buf.closed {
 		// Don't bother acking
 		return
 	}
 	buf.mx.RUnlock()
```

Reply:
eh, I guess the unlocking makes it messy -- current looks fine ignore me.

Reply:
Huh, almost identical code in the `submit` method. Two questions I don't know that answer to:

1) It seems like submit should possibly just use the same code you're using above, i.e.:

```
func (buf *receiveBuffer) submit(frame []byte) {
	buf.mx.RLock()
	closed := buf.closed
        buf.mx.RUnlock()
	if closed {
		return
	}
	buf.in <- frame
}
```

But does sending to `buf.in` need to be locked? If it does, then might `buf.ack` need to be locked as well?

2) Do `sendAck` and `submit` need to use the same lock?

Reply:
Looking at this more, it looks like `buf.mx` is purely protecting the closed state. If that's the case, then it seems like using the new code in `sendAck` could be the way to go.

Reply:
`buf.mx` is protecting `buf.in` as well. Once the receiver buffer is closed, `buf.in` is closed and we need to make sure not to send to it, otherwise we panic. `sendAck` is a little different because `buf.ack` doesn't get closed out from under us.